launch_ros/launch_ros/descriptions/node.py

```diff
@@ -0,0 +1,477 @@
+        launch.substitutions.TextSubstitution objects, and
+        `ros_specific_arguments` with any modifications made to it. If no
+        modifications are made, it should return
+        `[], ros_specific_arguments`.
+        """
+        return [], ros_specific_arguments
+
+
+class Node:
+    """Describes a ROS node."""
+
+    UNSPECIFIED_NODE_NAME = '<node_name_unspecified>'
+    UNSPECIFIED_NODE_NAMESPACE = '<node_namespace_unspecified>'
+
+    def __init__(
+        self, *,
+        node_name: Optional[SomeSubstitutionsType] = None,
+        node_namespace: Optional[SomeSubstitutionsType] = None,
+        original_node_name: Optional[SomeSubstitutionsType] = None,
+        parameters: Optional[SomeParameters] = None,
+        remappings: Optional[SomeRemapRules] = None,
+        traits: Optional[Iterable[NodeTrait]] = None,
+        **kwargs
+    ) -> None:
+        """
+        Construct a Node description.
+
+        If the node name is not given (or is None) then no name is passed to
+        the node on creation and instead the default name specified within the
+        code of the node is used instead.
+
+        The namespace can either be absolute (i.e. starts with /) or
+        relative.
+        If absolute, then nothing else is considered and this is passed
+        directly to the node to set the namespace.
+        If relative, the namespace in the 'ros_namespace' LaunchConfiguration
+        will be prepended to the given relative node namespace.
+        If no namespace is given, then the default namespace `/` is
+        assumed.
+
+        The parameters are passed as a list, with each element either a yaml
+        file that contains parameter rules (string or pathlib.Path to the full
+        path of the file), or a dictionary that specifies parameter rules.
+        Keys of the dictionary can be strings or an iterable of Substitutions
+        that will be expanded to a string.
+        Values in the dictionary can be strings, integers, floats, or tuples
+        of Substitutions that will be expanded to a string.
+        Additionally, values in the dictionary can be lists of the
+        aforementioned types, or another dictionary with the same properties.
+        A yaml file with the resulting parameters from the dictionary will be
+        written to a temporary file, the path to which will be passed to the
+        node.
+        Multiple dictionaries/files can be passed: each file path will be
+        passed in in order to the node (where the last definition of a
+        parameter takes effect).
+
+        :param: node_name the name of the node
+        :param: node_namespace the ROS namespace for this Node
+        :param: original_node_name the name of the node before remapping; if not specified,
+            remappings/parameters (including node name/namespace changes) may be applied
+            to all nodes which share a command line executable with this one
```

Comment:
@mlanting hmm, this is a bit clunky. Thinking out loud, perhaps `node_name` and `node_namespace` shouldn't be prescriptive but descriptive. That is, to be mandatory and expected to match what the actual node already defines, leaving name and namespace remapping to be done separately e.g.:

```python
Node(
   name = 'actual_name',
   namespace = 'actual_namespace'
   remappings = Node.naming(
        name = 'my_new_name', 
        namespace = 'my_new_namespace'
   ) + [('/chit', '/chat')],
)
```

CC @wjwwood @jacobperron for feedback.

Reply:
Either way it seems a bit brittle to me if we expect the name in launch to match the name in the code. You have to remember to update both places; and worse if the launch file is not maintained by the node author (e.g. in some end-users package).

However, I don't see another way around this problem besides introducing some far-fetched mechanism to query the coded node name from launch. I like @hidmic's suggestion since I think it will result in less hidden surprises for users (i.e. it avoids accidentally renaming a bunch of nodes in a process to the same name). But I'm also worried that being strict about providing the correct node name will be frustrating for users, who currently don't have to do this.

I'm torn.